util/merkle/src/snapshot/mod.rs

```diff
@@ -0,0 +1,338 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+mod chunk;
+mod compress;
+mod error;
+mod ordered_heap;
+
+use std::cmp::Ordering;
+
+use ccrypto::BLAKE_NULL_RLP;
+use cdb::HashDB;
+use primitives::H256;
+
+use self::chunk::{Chunk, RecoveredChunk, UnresolvedChunk};
+pub use self::compress::{ChunkCompressor, ChunkDecompressor};
+pub use self::error::Error;
+use self::ordered_heap::OrderedHeap;
+use crate::nibbleslice::NibbleSlice;
+
+const CHUNK_HEIGHT: usize = 3;
+const CHUNK_MAX_NODES: usize = 256; // 16 ^ (CHUNK_HEIGHT-1)
+
+/// Example:
+/// use codechain_merkle::snapshot::Restore;
+/// let mut rm = Restore::new(db, root);
+/// while let Some(root) = rm.next_to_feed() {
+///     let raw_chunk = request(block_hash, root)?;
+///     let chunk = raw_chunk.recover(root)?;
+///     rm.feed(chunk);
+/// }
+pub struct Restore<'a> {
+    db: &'a mut dyn HashDB,
+    pending: Option<ChunkPathPrefix>,
+    unresolved: OrderedHeap<DepthFirst<ChunkPathPrefix>>,
+}
+
+impl<'a> Restore<'a> {
+    pub fn new(db: &'a mut dyn HashDB, merkle_root: H256) -> Self {
+        let mut result = Restore {
+            db,
+            pending: None,
+            unresolved: OrderedHeap::new(),
+        };
+        if merkle_root != BLAKE_NULL_RLP {
+            result.unresolved.push(ChunkPathPrefix::new(merkle_root).into());
+        }
+        result
+    }
+
+    pub fn feed(&mut self, chunk: RecoveredChunk) {
+        let pending_path = self.pending.take().expect("feed() should be called after next()");
+        assert_eq!(pending_path.chunk_root, chunk.root, "Unexpected chunk");
+
+        // Pour nodes into the DB
+        for (_, value) in chunk.nodes {
+            self.db.insert(&value);
+        }
+
+        // Extend search paths
+        for unresolved in chunk.unresolved_chunks {
+            self.unresolved.push(pending_path.with_unresolved_chunk(&unresolved).into());
+        }
+
+        self.pending = None;
+    }
+
+    pub fn next_to_feed(&mut self) -> Option<H256> {
+        if let Some(path) = self.unresolved.pop() {
+            assert!(self.pending.is_none(), "Previous feed() was failed");
+            let chunk_root = path.chunk_root;
+            self.pending = Some(path.0);
+
+            Some(chunk_root)
+        } else {
+            None
+        }
+    }
+}
+
+/// Example:
+/// use std::fs::File;
+/// use codechain_merkle::snapshot::Snapshot;
+///
+/// for chunk in Snapshot::from_hashdb(db, root) {
+///     let mut file = File::create(format!("{}/{}", block_id, chunk.root))?;
+///     let mut compressor = ChunkCompressor::new(&mut file);
+///     compressor.compress(chunk);
+/// }
+pub struct Snapshot<'a> {
+    db: &'a dyn HashDB,
+    remaining: OrderedHeap<DepthFirst<ChunkPathPrefix>>,
```

Comment:
Why should it be ordered?

Reply:
Rust's `BinaryHeap` doesn't guarantee the insertion order between objects  `if let Ordering::Equal = partial_cmp(a, b)`. I wanted to make them ordered.

Reply:
So why do you want to guarantee the insertion order when the value is equal?

Reply:
It is easier to reason about when debugging.

Reply:
Okay. Thank you for your comment.